packages/plugin/src/migration.ts

```diff
@@ -1,134 +1,89 @@
-        }
-        return type.deserialize(serialized, helpers)
-      },
-      serialize(
-        deserialized: T,
-        helpers: StoreSerializeHelpers
-      ): Versionized<S> {
-        return {
-          __version__: 0,
-          value: type.serialize(deserialized, helpers)
-        }
-      },
-      recursiveMigrate(previousState: S | Versionized<S>): S {
-        if (isVersionized<S>(previousState, 0)) {
-          return previousState.value
-        }
-        return previousState
+  type: StateType<S, T, R>
+): MigratableStateType<S, S, S, T, R> {
+  return migrate<S, S, S, T, R, S, T, R>(
+    state => {
+      if (isVersionized(state, 0)) {
+        return state.value
       }
-    }
+      return state
+    },
+    type,
+    0,
+    state => state
   )
-
-  return Object.assign(state, { migrate: nextMigrate })
-
-  function nextMigrate<S1, T1, R1>(
-    nextNextType: StateDescriptor<S1, T1, R1>,
-    f: (previousState: S) => S1
-  ) {
-    return migrate(state, nextNextType, 0, f)
-  }
 }
 
 function migrate<InitialS, AllS, S, T, R, S1, T1, R1>(
-  currentType: AbstractMigratableStateDescriptor<InitialS, AllS, S, T, R>,
-  nextType: StateDescriptor<S1, T1, R1>,
-  currentVersion: number,
+  recursiveMigrate: (previousState: InitialS | Versionized<AllS>) => S,
+  nextType: StateType<S1, T1, R1>,
+  nextVersion: number,
   f: (previousState: S) => S1
-): MigratableStateDescriptor<InitialS, AllS | S1, S1, T1, R1> {
-  const nextVersion = currentVersion + 1
-  const state: AbstractMigratableStateDescriptor<
-    InitialS,
-    AllS | S1,
-    S1,
-    T1,
-    R1
-  > = Object.assign(
-    function(
+): MigratableStateType<InitialS, AllS | S1, S1, T1, R1> {
+  // eslint-disable-next-line @typescript-eslint/no-extraneous-class
+  class MigrateType {
+    constructor(
       value: T1,
       onChange: (
         updater: (oldValue: T1, helpers: StoreDeserializeHelpers) => T1
       ) => void,
       pluginProps?: unknown
-    ): R1 {
-      return nextType(value, onChange, pluginProps)
-    },
-    {
-      createInitialState(helpers: StoreDeserializeHelpers) {
-        return nextType.createInitialState(helpers)
-      },
-      deserialize(
-        serialized: InitialS | Versionized<AllS | S1>,
-        helpers: StoreDeserializeHelpers
-      ): T1 {
-        if (isVersionized<S1>(serialized, nextVersion)) {
-          return nextType.deserialize(serialized.value, helpers)
-        }
-        const s = serialized as InitialS | Versionized<AllS>
-        return nextType.deserialize(f(currentType.recursiveMigrate(s)), helpers)
-      },
-      serialize(
-        deserialized: T1,
-        helpers: StoreSerializeHelpers
-      ): Versionized<S1> {
-        return {
-          __version__: nextVersion,
-          value: nextType.serialize(deserialized, helpers)
-        }
-      },
-      recursiveMigrate
+    ) {
+      Object.assign(this, new nextType(value, onChange, pluginProps))
```

Comment:
That's a little bit weird IMHO. Another possible solution is to provide the underlying state type as a property, e.g.
```
const state = migratable(state1).migrate(state2, f)

function C(props) {
  const state = props.state.unwrap()
}
```

Reply:
Or something like
```
const state = migratable(state1).migrate(state2, f).unwrap() /* no longer able to apply migrate after unwrap */

function C(props) {
  const state = props.state
}```

Reply:
So the problem is, that we can't return an instance of another class from constructor call, right? What if we change it to `static create(value, onChange)` instead of `constructor (value, onChange)` to ? And then use the StateTypes as `plugin.state.create(state, onChange)` instead of `new plugin.state(state, onChange)`?

Reply:
Nice idea :)